simexpal/base.py

```diff
@@ -1128,26 +1159,71 @@ def output_file_path(self, ext):
 		return os.path.join(self.experiment.output_subdir,
 				get_output_file_name(ext, self.instance.shortname, self.repetition))
 
-	def get_status(self):
-		if os.access(self.output_file_path('status'), os.F_OK):
-			with open(self.output_file_path('status'), "r") as f:
+	def _update_status_cache_dict(self):
+
+		# Determine status cache entry.
+		status, last_mod = Status.NOT_SUBMITTED, None
+		try:
+			with open(self.output_file_path('status'), 'r') as f:
 				status_dict = yaml.load(f, Loader=YmlLoader)
+				last_mod = os.fstat(f.fileno()).st_mtime
 
 			if status_dict['timeout']:
-				return Status.TIMEOUT
+				status = Status.TIMEOUT
 			elif status_dict['signal']:
-				return Status.KILLED
+				status = Status.KILLED
 			elif status_dict['status'] > 0:
-				return Status.FAILED
-			return Status.FINISHED
-		elif os.access(self.output_file_path('out'), os.F_OK):
-			return Status.STARTED
-		elif os.access(self.aux_file_path('run'), os.F_OK):
-			return Status.SUBMITTED
-		elif os.access(self.aux_file_path('lock'), os.F_OK):
-			return Status.IN_SUBMISSION
-
-		return Status.NOT_SUBMITTED
+				status = Status.FAILED
+			else:
+				status = Status.FINISHED
+		except FileNotFoundError:
+			if os.access(self.output_file_path('out'), os.F_OK):
+				status, last_mod = Status.STARTED, os.stat(self.output_file_path('out')).st_mtime
+			elif os.access(self.aux_file_path('run'), os.F_OK):
+				status, last_mod = Status.SUBMITTED, os.stat(self.aux_file_path('run')).st_mtime
+			elif os.access(self.aux_file_path('lock'), os.F_OK):
+				status, last_mod = Status.IN_SUBMISSION, os.stat(self.aux_file_path('lock')).st_mtime
+
+		internal_name = self.internal_name
+		if internal_name not in self._cfg.status_cache_dict:
+			self._cfg.status_cache_dict[internal_name] = {
+				self.instance.shortname: {'status': status, 'last_mod': last_mod}}
+		else:
+			self._cfg.status_cache_dict[internal_name][self.instance.shortname] = {
+				'status': status, 'last_mod': last_mod}
+
+		return status
+
+	def purge_status_cache_dict(self):
+		try:
+			del self._cfg.status_cache_dict[self.internal_name][self.instance.shortname]
+		except KeyError:
+			pass
+
+	def get_status(self):
+
+		last_mod = None
+		try:
+			last_mod = os.stat(self.output_file_path('status')).st_mtime
+		except FileNotFoundError:
+			try:
+				last_mod = os.stat(self.output_file_path('out')).st_mtime
+			except FileNotFoundError:
+				try:
+					last_mod = os.stat(self.aux_file_path('run')).st_mtime
+				except FileNotFoundError:
+					try:
+						last_mod = os.stat(self.aux_file_path('lock')).st_mtime
+					except FileNotFoundError:
+						pass
+
+		internal_name = self.internal_name
+		if internal_name in self._cfg.status_cache_dict:
+			if self.instance.shortname in self._cfg.status_cache_dict[internal_name]:
+				cache_entry = self._cfg.status_cache_dict[internal_name][self.instance.shortname]
+				if last_mod == cache_entry["last_mod"]:
```

Comment:
Small optimization: we only need to compute `last_mod` here, i.e., before the `if`. This should make it somewhat faster to add new entries to the cache.

Reply:
This code could potentially break if the modification time of two files is exactly identical (e.g., `.status` was not found yet but has the same timestamp as `.lock`). That can probably not happen on modern file systems with sub-second precision but it might be worth to add a comment here; we do not need to fix it now.

Reply:
> Small optimization: we only need to compute `last_mod` here, i.e., before the `if`. This should make it somewhat faster to add new entries to the cache.

Done.



> This code could potentially break if the modification time of two files is exactly identical (e.g., `.status` was not found yet but has the same timestamp as `.lock`). That can probably not happen on modern file systems with sub-second precision but it might be worth to add a comment here; we do not need to fix it now.

To fix it, we could also save the file extension of the file that produced this timestamp. Although the case above should not really occur, except when the user manually deletes status related files **and** those files have the same timestamp.

Reply:
True. We probably do not even need to store additional info, we can reconstruct from the cached status which file we read (and compare that to the last_mod file).

Reply:
I have added another check so that the above scenario can not happen.